File: src/main/java/de/twometer/amongus/gui/TaskPage.java
package de.twometer.amongus.gui;

import de.twometer.amongus.core.AmongUs;
import de.twometer.amongus.model.Location;
import de.twometer.amongus.model.PlayerTask;
import de.twometer.amongus.model.TaskType;
import de.twometer.amongus.net.NetMessage;
import de.twometer.amongus.util.RandomUtil;

import javax.sound.sampled.FloatControl;
import java.util.Random;

public class TaskPage extends BasePage {

    private final Location location;
    private final TaskType taskType;

    private final PlayerTask task;

    public TaskPage(Location location, TaskType taskType) {
        super("Tasks/" + taskType.name() + ".html");
        this.location = location;
        this.taskType = taskType;
        this.task = AmongUs.get().getSession().getMyself().findRunningTaskByStage(location, taskType);
    }

    @Override
    public void onDomReady() {
        super.onDomReady();
        if (taskType == TaskType.DataTransfer && location != Location.Admin) {
            context.call("ConfigureTaskUI", false, TaskFormatter.formatLocation(location));
        } else if (taskType == TaskType.DivertPower) {
            var destination = task.getLastStage().getLocation().name().toLowerCase();
            context.call("setSliderActive", destination);
        } else if (taskType == TaskType.Scan) {
            var player = amongUs.getSession().getMyself();
            var id = Integer.toHexString(player.id ^ hashCode()).substring(0, 6).toUpperCase();
            var height = RandomUtil.nextInt(55, 70);
            var weight = RandomUtil.nextInt(35, 55);
            var bt = new String[]{"A", "B", "AB", "0"}[RandomUtil.nextInt(0, 4)];
            var state = player.alive ? "HEALTHY" : "GHOST";
            context.call("setProperties", id, player.color.name(), height, weight, bt, state);
        } else if (taskType == TaskType.InspectSamples && task.isTimerEnded()) {
            context.call("setAnomaly", RandomUtil.nextInt(0, 5));
        }
    }

    public void beginTimeout() {
        task.startTimer(55);
    }

    public void taskComplete() {
        task.advance();
        if (task.isCompleted()) {
            AmongUs.get().getSoundFX().play("TaskComplete.ogg");
        } else {
            AmongUs.get().getSoundFX().play("TaskProgress.ogg");
        }
        AmongUs.get().getClient().sendMessage(new NetMessage.CompleteTaskStage());
        close();
    }

    public void close() {
        AmongUs.get().getScheduler().runLater(800, () -> {
            if (AmongUs.get().getStateController().isRunning())
                this.goBack();
        });
    }

    @Override
    protected boolean escapeGoesBack() {
        return true;
    }
}

File: src/main/kotlin/de/twometer/amogus/gui/TaskPage.kt
package de.twometer.amogus.gui

import de.twometer.amogus.client.AmongUsClient
import de.twometer.amogus.client.TaskFormatter
import de.twometer.amogus.client.TaskGameObject
import de.twometer.amogus.model.Location
import de.twometer.amogus.model.PlayerState
import de.twometer.amogus.model.PlayerTask
import de.twometer.amogus.model.TaskType
import de.twometer.amogus.net.CompleteTaskStage
import de.twometer.neko.audio.SoundEngine
import kotlin.math.floor
import kotlin.random.Random

class TaskPage(private val taskObject: TaskGameObject, private val task: PlayerTask) : BasePage("Tasks/${taskObject.taskType}.html") {

    override fun onLoaded() {
        val type = taskObject.taskType
        val location = taskObject.location
        if (type == TaskType.DataTransfer) {
            call("ConfigureTaskUI", location == Location.Admin, TaskFormatter.formatLocation(location))
        } else if (type == TaskType.DivertPower) {
            val destination = task.lastStage.location.name.lowercase()
            call("setSliderActive", destination)
        } else if (type == TaskType.Scan) {
            val player = AmongUsClient.session!!.myself
            val hash = player.username.hashCode()
            val random = Random(hash)
            val id = Integer.toHexString(hash).uppercase()
            val height = random.nextInt(15) + 55
            val weight = random.nextInt(20) + 35
            val blood = arrayOf("A", "B", "AB", "0")[random.nextInt(4)]
            val state = if (player.state == PlayerState.Ghost) "GHOST" else "ALIVE"
            call("setProperties", id, player.color.name, height, weight, blood, state)
        } else if (type == TaskType.InspectSamples && task.isTimerEnded) {
            call("setAnomaly", floor(Math.random() * 5))
        }
    }

    fun beginTimeout() {
        task.startTimer(55)
    }

    fun taskComplete() {
        task.advance()
        if (task.isCompleted)
            SoundEngine.play("TaskComplete.ogg")
        else
            SoundEngine.play("TaskProgress.ogg")
        AmongUsClient.send(CompleteTaskStage())
        AmongUsClient.mainScheduler.runLater(800) {
            close()
        }
    }

}